File: packages/ops-handoff-core/tests/test_ops_handoff_core.py
#!/usr/bin/env python3
"""Static behavior tests for ops-handoff-core."""

from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path


def run(cmd: list[str], *, expect: int = 0) -> subprocess.CompletedProcess[str]:
    proc = subprocess.run(cmd, text=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    if proc.returncode != expect:
        raise AssertionError(
            f"unexpected exit {proc.returncode}, expected {expect}: {' '.join(cmd)}\n"
            f"stdout:\n{proc.stdout}\nstderr:\n{proc.stderr}"
        )
    return proc


def main(argv: list[str]) -> int:
    package_dir = Path(argv[0]).resolve()
    out_root = Path(argv[1]).resolve()
    bin_path = package_dir / "bin" / "ops-handoff-core.py"
    fixtures = package_dir / "tests" / "fixtures"
    handoff = out_root / "handoff"
    out_root.mkdir(parents=True, exist_ok=True)

    generate = run([
        sys.executable,
        str(bin_path),
        "generate",
        "--role-catalog",
        str(fixtures / "role-catalog.md"),
        "--topology",
        str(fixtures / "organization-topology.a2ui.jsonl"),
        "--command-board",
        str(fixtures / "command-board.a2ui.jsonl"),
        "--request",
        str(fixtures / "REQUEST.md"),
        "--source-manifest",
        str(fixtures / "source-manifest.json"),
        "--runtime-manifest",
        str(fixtures / "runtime-manifest.json"),
        "--merge-target",
        str(fixtures / "merge-target.json"),
        "--thread-roster",
        str(fixtures / "thread-roster.json"),
        "--out-dir",
        str(handoff),
        "--json",
    ])
    generated = json.loads(generate.stdout)
    assert generated["status"] == "handoff-generated"

    validate = run([
        sys.executable,
        str(bin_path),
        "validate",
        "--handoff-dir",
        str(handoff),
        "--no-role-body-sentinel",
        "FULL_ROLE_CATALOG_BODY_SENTINEL",
    ])
    valid = json.loads(validate.stdout)
    assert valid["status"] == "handoff-valid"

    manifest = json.loads((handoff / "HANDOFF_MANIFEST.json").read_text(encoding="utf-8"))
    assert manifest["handoffId"].startswith("handoff:")
    assert manifest["handoffId"] != "handoff:ops-handoff-core-proof"
    assert manifest["state"]["current"] == "handoff-created"
    assert manifest["state"]["terminal"] is False
    assert manifest["approvalBoundary"]["transportReadbackIsApproval"] is False
    assert manifest["approvalBoundary"]["semanticApproval"] is False
    assert manifest["approvalBoundary"]["completionApproval"] is False
    assert {row["threadFunction"] for row in manifest["threads"]} == {
        "impl-work",
        "impl-review",
        "merge-work",
        "merge-review",
    }

    thread_text = "\n".join(path.read_text(encoding="utf-8") for path in (handoff / "THREADS").glob("*/*.md"))
    assert "project-source-put" not in thread_text
    assert "project-thread-create" not in thread_text
    assert "project-artifact-fetch" not in thread_text
    assert "FULL_ROLE_CATALOG_BODY_SENTINEL" not in thread_text

    missing = run([
        sys.executable,
        str(bin_path),
        "generate",
        "--role-catalog",
        str(fixtures / "role-catalog.md"),
        "--topology",
        str(fixtures / "organization-topology.a2ui.jsonl"),
        "--command-board",
        str(fixtures / "command-board.a2ui.jsonl"),
        "--request",
        str(fixtures / "REQUEST.md"),
        "--source-manifest",
        str(fixtures / "source-manifest.json"),
        "--runtime-manifest",
        str(fixtures / "runtime-manifest.json"),
        "--merge-target",
        str(fixtures / "merge-target.json"),
        "--out-dir",
        str(out_root / "missing-roster"),
        "--json",
    ], expect=2)
    missing_result = json.loads(missing.stdout)
    assert missing_result["status"] == "missing-required-input"

    artifact = out_root / "artifact.txt"
    run_report = out_root / "RUN_REPORT.md"
    verdict = out_root / "verdict.txt"
    claim_path = out_root / "claim.jsonl"
    artifact.write_text("artifact\n", encoding="utf-8")
    run_report.write_text("# run report\nok\n", encoding="utf-8")
    verdict.write_text("merge-review-pass\nok\n", encoding="utf-8")
    imported = run([
        sys.executable,
        str(bin_path),
        "import-result",
        "--thread-function",
        "merge-review",
        "--artifact",
        str(artifact),
        "--run-report",
        str(run_report),
        "--verdict-file",
        str(verdict),
        "--claim-path",
        str(claim_path),
        "--json",
    ])
    imported_doc = json.loads(imported.stdout)
    assert imported_doc["status"] == "handoff-result-imported"
    assert imported_doc["localizerApproval"] is False
    assert claim_path.is_file()
    claim = json.loads(claim_path.read_text(encoding="utf-8").splitlines()[-1])
    assert claim["approvalBoundary"]["localizerApproval"] is False
    assert claim["artifacts"][0]["sha256"]
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: packages/ops-handoff-core/tests/test_ops_handoff_core.mjs
#!/usr/bin/env node
// Static behavior tests for ops-handoff-core (Node port of test_ops_handoff_core.py).

import fs from "node:fs";
import path from "node:path";
import process from "node:process";
import { spawnSync } from "node:child_process";

process.on("unhandledRejection", (e) => {
  console.error(e);
  process.exit(1);
});

function assert(cond, message) {
  if (!cond) throw new Error("AssertionError: " + (message || ""));
}

function run(cmd, expect = 0) {
  const proc = spawnSync(cmd[0], cmd.slice(1), { encoding: "utf-8" });
  const code = proc.status;
  if (code !== expect) {
    throw new Error(
      `unexpected exit ${code}, expected ${expect}: ${cmd.join(" ")}\n` +
        `stdout:\n${proc.stdout}\nstderr:\n${proc.stderr}`,
    );
  }
  return proc;
}

function main(argv) {
  const packageDir = path.resolve(argv[0]);
  const outRoot = path.resolve(argv[1]);
  const binPath = path.join(packageDir, "bin", "ops-handoff-core.mjs");
  const fixtures = path.join(packageDir, "tests", "fixtures");
  const handoff = path.join(outRoot, "handoff");
  fs.mkdirSync(outRoot, { recursive: true });

  const node = process.execPath;

  const generate = run([
    node,
    binPath,
    "generate",
    "--role-catalog",
    path.join(fixtures, "role-catalog.md"),
    "--topology",
    path.join(fixtures, "organization-topology.a2ui.jsonl"),
    "--command-board",
    path.join(fixtures, "command-board.a2ui.jsonl"),
    "--request",
    path.join(fixtures, "REQUEST.md"),
    "--source-manifest",
    path.join(fixtures, "source-manifest.json"),
    "--runtime-manifest",
    path.join(fixtures, "runtime-manifest.json"),
    "--merge-target",
    path.join(fixtures, "merge-target.json"),
    "--thread-roster",
    path.join(fixtures, "thread-roster.json"),
    "--out-dir",
    handoff,
    "--json",
  ]);
  const generated = JSON.parse(generate.stdout);
  assert(generated.status === "handoff-generated", "generate status");

  const validate = run([
    node,
    binPath,
    "validate",
    "--handoff-dir",
    handoff,
    "--no-role-body-sentinel",
    "FULL_ROLE_CATALOG_BODY_SENTINEL",
  ]);
  const valid = JSON.parse(validate.stdout);
  assert(valid.status === "handoff-valid", "validate status");

  const manifest = JSON.parse(fs.readFileSync(path.join(handoff, "HANDOFF_MANIFEST.json"), "utf-8"));
  assert(manifest.handoffId.startsWith("handoff:"), "handoffId prefix");
  assert(manifest.handoffId !== "handoff:ops-handoff-core-proof", "handoffId not stub");
  assert(manifest.state.current === "handoff-created", "state current");
  assert(manifest.state.terminal === false, "state terminal");
  assert(manifest.approvalBoundary.transportReadbackIsApproval === false, "transportReadbackIsApproval");
  assert(manifest.approvalBoundary.semanticApproval === false, "semanticApproval");
  assert(manifest.approvalBoundary.completionApproval === false, "completionApproval");
  const fns = new Set(manifest.threads.map((row) => row.threadFunction));
  const want = new Set(["impl-work", "impl-review", "merge-work", "merge-review"]);
  assert(fns.size === want.size && [...want].every((f) => fns.has(f)), "thread functions set");

  const threadsRoot = path.join(handoff, "THREADS");
  const mdFiles = [];
  for (const dir of fs.readdirSync(threadsRoot).sort()) {
    const full = path.join(threadsRoot, dir);
    if (fs.statSync(full).isDirectory()) {
      for (const f of fs.readdirSync(full).sort()) {
        if (f.endsWith(".md")) mdFiles.push(path.join(full, f));
      }
    }
  }
  const threadText = mdFiles.map((p) => fs.readFileSync(p, "utf-8")).join("\n");
  assert(!threadText.includes("project-source-put"), "no project-source-put");
  assert(!threadText.includes("project-thread-create"), "no project-thread-create");
  assert(!threadText.includes("project-artifact-fetch"), "no project-artifact-fetch");
  assert(!threadText.includes("FULL_ROLE_CATALOG_BODY_SENTINEL"), "no sentinel");

  const missing = run(
    [
      node,
      binPath,
      "generate",
      "--role-catalog",
      path.join(fixtures, "role-catalog.md"),
      "--topology",
      path.join(fixtures, "organization-topology.a2ui.jsonl"),
      "--command-board",
      path.join(fixtures, "command-board.a2ui.jsonl"),
      "--request",
      path.join(fixtures, "REQUEST.md"),
      "--source-manifest",
      path.join(fixtures, "source-manifest.json"),
      "--runtime-manifest",
      path.join(fixtures, "runtime-manifest.json"),
      "--merge-target",
      path.join(fixtures, "merge-target.json"),
      "--out-dir",
      path.join(outRoot, "missing-roster"),
      "--json",
    ],
    2,
  );
  const missingResult = JSON.parse(missing.stdout);
  assert(missingResult.status === "missing-required-input", "missing-required-input");

  const artifact = path.join(outRoot, "artifact.txt");
  const runReport = path.join(outRoot, "RUN_REPORT.md");
  const verdict = path.join(outRoot, "verdict.txt");
  const claimPath = path.join(outRoot, "claim.jsonl");
  fs.writeFileSync(artifact, "artifact\n", "utf-8");
  fs.writeFileSync(runReport, "# run report\nok\n", "utf-8");
  fs.writeFileSync(verdict, "merge-review-pass\nok\n", "utf-8");
  const imported = run([
    node,
    binPath,
    "import-result",
    "--thread-function",
    "merge-review",
    "--artifact",
    artifact,
    "--run-report",
    runReport,
    "--verdict-file",
    verdict,
    "--claim-path",
    claimPath,
    "--json",
  ]);
  const importedDoc = JSON.parse(imported.stdout);
  assert(importedDoc.status === "handoff-result-imported", "import status");
  assert(importedDoc.localizerApproval === false, "localizerApproval");
  assert(fs.statSync(claimPath).isFile(), "claim file");
  const lines = fs.readFileSync(claimPath, "utf-8").split(/\r\n|\r|\n/).filter((l) => l !== "");
  const claim = JSON.parse(lines[lines.length - 1]);
  assert(claim.approvalBoundary.localizerApproval === false, "claim localizerApproval");
  assert(claim.artifacts[0].sha256, "artifact sha256");
  return 0;
}

process.exit(main(process.argv.slice(2)));
